ledger/simulation/tracer_test.go

```diff
@@ -0,0 +1,257 @@
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{0, 2}},
+				{action: mocktracer.BeforeTxnGroupEvent},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{0, 2, 0}},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{0, 2, 0}},
+				{action: mocktracer.AfterTxnGroupEvent},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{0, 2}},
+				{action: mocktracer.AfterTxnGroupEvent},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{0}},
+				{action: mocktracer.AfterTxnGroupEvent},
+			},
+		},
+		{
+			name: "second txn with deep inners",
+			timeline: []step{
+				{action: mocktracer.BeforeTxnGroupEvent},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{0}},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{0}},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{1}},
+				{action: mocktracer.BeforeTxnGroupEvent},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{1, 0}},
+				{action: mocktracer.BeforeTxnGroupEvent},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{1, 0, 0}},
+				{action: mocktracer.BeforeTxnGroupEvent},
+				{action: mocktracer.BeforeTxnEvent, expectedPath: TxnPath{1, 0, 0, 0}},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{1, 0, 0, 0}},
+				{action: mocktracer.AfterTxnGroupEvent},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{1, 0, 0}},
+				{action: mocktracer.AfterTxnGroupEvent},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{1, 0}},
+				{action: mocktracer.AfterTxnGroupEvent},
+				{action: mocktracer.AfterTxnEvent, expectedPath: TxnPath{1}},
+				{action: mocktracer.AfterTxnGroupEvent},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("%s", tc.name), func(t *testing.T) {
+			t.Parallel()
+			var tracer cursorEvalTracer
+
+			// These don't matter so they can be anything
+			ep := logic.EvalParams{}
+			groupIndex := 0
+
+			for i, step := range tc.timeline {
+				switch step.action {
+				case mocktracer.BeforeTxnEvent:
+					tracer.BeforeTxn(&ep, groupIndex)
+				case mocktracer.AfterTxnEvent:
+					tracer.AfterTxn(&ep, groupIndex, transactions.ApplyData{}, nil)
+				case mocktracer.BeforeTxnGroupEvent:
+					tracer.BeforeTxnGroup(&ep)
+				case mocktracer.AfterTxnGroupEvent:
+					tracer.AfterTxnGroup(&ep, nil)
+				default:
+					t.Fatalf("unexpected timeline hook: %v", step.action)
+				}
+				if step.expectedPath != nil {
+					switch step.action {
+					case mocktracer.BeforeTxnGroupEvent, mocktracer.AfterTxnGroupEvent:
+						t.Fatalf("Path is unspecified for hook: %v", step.action)
+					}
+					require.Equalf(t, step.expectedPath, tracer.absolutePath(), "step index %d (action %v), tracer: %#v", i, step.action, tracer)
+				}
+			}
+
+			if tc.expectedPathAtEnd != nil {
+				require.Equal(t, tc.expectedPathAtEnd, tracer.absolutePath())
+			}
+		})
+	}
+}
```

Comment:
No before/after program or before/after opcode test?

Reply:
This test is for `cursorEvalTracer`, whose job is to maintain the "path" of the current transaction being evaluated. Individual opcode events aren't significant for that, so they aren't tested here